src/Architectures.jl

```diff
@@ -25,23 +26,34 @@ variable `JULIA_NUM_THREADS` is set.
 struct CPU <: AbstractArchitecture end
 
 """
-    GPU <: AbstractArchitecture
+    GPU{D} <: AbstractArchitecture
 
-Run Oceananigans on a single NVIDIA CUDA GPU.
+Run Oceananigans on a single Nvidia CUDA or AMD ROCm GPU.
 """
-struct GPU <: AbstractArchitecture end
+struct GPU{D} <: AbstractArchitecture
+    device :: D
+end
+
+const CUDAGPU = GPU{<:CUDA.CUDABackend}
+const ROCmGPU = GPU{<:AMDGPU.ROCBackend}
+
+GPU() = has_cuda() ? GPU(CUDA.CUDABackend()) : GPU(AMDGPU.ROCBackend())
```

Comment:
this implies that if there is no CUDA-enabled GPU then `GPU()` will default to `GPU(AMDGPU.ROCBackend())`.  But what if there is no AMD GPU either?